pandas/_libs/lib.pyx

```diff
@@ -598,6 +598,43 @@ def maybe_booleans_to_slice(ndarray[uint8_t, ndim=1] mask):
         return slice(start, end)
 
 
+def invert_object_array(ndarray values):
+    # GH#54569
+    cdef:
+        Py_ssize_t i, N = values.size
+        object val, res_val
+        ndarray result = cnp.PyArray_EMPTY(values.ndim, values.shape, cnp.NPY_OBJECT, 0)
+        object[::1] res_flat = result.ravel()  # should NOT be a copy
+        cnp.flatiter it = cnp.PyArray_IterNew(values)
+
+    for i in range(N):
+
+        # Analogous to: val = values[i]
+        val = PyArray_GETITEM(values, PyArray_ITER_DATA(it))
```

Comment:
Does this make any difference compared to the normal array access? I would be surprised if so - the annotate functionality is a good way to check

Reply:
this is the pattern we use to handle arbitrary-ndim ndarrays